### ISSUE TITLE
LeNet-5 FC3 float16 forward pass precision issue

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ProjectOdyssey#2703]
## Summary

The `test_fc3_forward_float16` test in `tests/models/test_lenet5_layers.mojo` crashes due to float16 precision limitations.

## Details

- **Test**: `test_fc3_forward_float16`
- **Layer**: FC3 (84 inputs -> 10 outputs)
- **Issue**: 840 multiplications per output element causes float16 overflow/precision loss

## Root Cause

Float16 has limited precision (~3.3 decimal digits) and range (max ~65504). Large accumulations can exceed these limits.

## Workaround

Test is marked with `FIXME(#2703)` and skipped.

## Fix Options

1. Use float32 for accumulation even with float16 tensors
2. Use Kahan summation for better precision
3. Accept float16 has inherent precision limitations for large accumulations

**Comment by mvillmow** [comment]
## Detailed Implementation Plan

Now I have a complete picture. Let me create the comprehensive implementation plan:

---

# Implementation Plan: LeNet-5 FC3 Float16 Forward Pass Precision Issue (#2703)

## Summary

The `test_fc3_forward_float16` test crashes due to float16 precision limitations when performing 84 multiplications per output element in the FC3 layer (84→10). The root cause is naive float16 accumulation in the matrix multiplication implementation (`shared/core/matrix.mojo`). This plan proposes implementing **float32 accumulation with float16 storage** as the recommended solution, with an option to add Kahan summation as an alternative for extreme precision requirements.

## Problem Analysis

### Root Cause

The matrix multiplication implementation in `shared/core/matrix.mojo` performs accumulation in the same dtype as the input tensors:

```mojo
fn _matmul_2d_1d_impl[dtype: DType](...):
    for i in range(m):
        var sum_val: Scalar[dtype] = 0  # ← Accumulator is float16
        for j in range(k):
            sum_val += a_ptr[i * k + j] * b_ptr[j]  # ← All ops in float16
        out_ptr[i] = sum_val
```

For float16 with 84 multiplications:
- Each operation introduces ~4.88×10⁻⁴ relative error (machine epsilon)
- After 84 operations: accumulated error ≈ √84 × ε ≈ 0.44% relative error
- This exceeds tolerance for small logit values in classification tasks

### Why FC3 Specifically Fails

1. **Small output dimension (10)** → small logit values (typically -5 to 5)
2. **Absolute error matters** → 0.1 error can flip softmax rankings
3. **Test values expose precision loss** → using ones reveals systematic error accumulation

### Fix Options (from Issue)

1. **Use float32 for accumulation even with float16 tensors** (Recommended)
2. **Use Kahan summation for better precision** (Alternative)
3. **Accept float16 has inherent precision limitations** (Document/Skip)

## Recommended Approach: Option 1 - Float32 Accumulation

This is the industry-standard approach used by PyTorch, TensorFlow, and CUDA libraries. It provides:
- ~10⁷ times better precision in accumulation
- Minimal performance overhead (accumulator is register, not memory)
- No change to input/output tensor types

## Step-by-Step Implementation Tasks

### Phase 1: Core Matrix Multiplication Changes

#### Task 1.1: Add Float32 Accumulation to 2D×1D Matmul
**File**: `/home/mvillmow/ml-odyssey/shared/core/matrix.mojo`
**Lines**: 22-34

Create a new implementation that uses float32 accumulation for reduced-precision dtypes:

```mojo
fn _matmul_2d_1d_impl_mixed_precision[
    dtype: DType
](result: ExTensor, a: ExTensor, b: ExTensor, m: Int, k: Int):
    """2D @ 1D matmul with float32 accumulation for reduced precision types."""
    var a_ptr = a._data.bitcast[Scalar[dtype]]()
    var b_ptr = b._data.bitcast[Scalar[dtype]]()
    var out_ptr = result._data.bitcast[Scalar[dtype]]()

    for i in range(m):
        # Use Float32 accumulator regardless of input dtype
        var sum_val: Float32 = 0.0
        for j in range(k):
            # Convert to Float32 for accumulation
            sum_val += Float32(a_ptr[i * k + j]) * Float32(b_ptr[j])
        # Convert back to output dtype
        out_ptr[i] = Scalar[dtype](sum_val)
```

#### Task 1.2: Add Float32 Accumulation to 2D×2D Matmul
**File**: `/home/mvillmow/ml-odyssey/shared/core/matrix.mojo`
**Lines**: 90-111

Apply same pattern to 2D×2D matrix multiplication.

#### Task 1.3: Add Float32 Accumulation to Batched Matmul
**File**: `/home/mvillmow/ml-odyssey/shared/core/matrix.mojo`
**Lines**: 137-170

Apply same pattern to batched matrix multiplication.

#### Task 1.4: Update Dispatch Functions
**File**: `/home/mvillmow/ml-odyssey/shared/core/matrix.mojo`
**Lines**: 37-54, 113-135, 172-253

Modify dispatch functions to use mixed-precision implementations for float16/bfloat16:

```mojo
fn _dispatch_matmul_2d_1d(
    result: ExTensor, a: ExTensor, b: ExTensor, m: Int, k: Int
) raises:
    """Runtime dispatch for 2D @ 1D matmul."""
    var dt = a.dtype()
    if dt == DType.float16:
        _matmul_2d_1d_impl_mixed_precision[DType.float16](result, a, b, m, k)
    elif dt == DType.bfloat16:
        _matmul_2d_1d_impl_mixed_precision[DType.bfloat16](result, a, b, m, k)
    elif dt == DType.float32:
        _matmul_2d_1d_impl[DType.float32](result, a, b, m, k)
    # ... existing cases for float64, int32, int64
```

#### Task 1.5: Add Float32 Accumulation to Dot Product
**File**: `/home/mvillmow/ml-odyssey/shared/core/matrix.mojo`
**Lines**: 332-363

Apply same pattern to dot product implementation.

### Phase 2: Reduction Operations (Same Issue Pattern)

#### Task 2.1: Update Sum Operations
**File**: `/home/mvillmow/ml-odyssey/shared/core/reduction.mojo`

Apply float32 accumulation pattern to sum operations:

```mojo
fn _sum_all_impl_mixed_precision[dtype: DType](result: ExTensor, tensor: ExTensor, numel: Int):
    """Sum over all elements with float32 accumulation."""
    var in_ptr = tensor._data.bitcast[Scalar[dtype]]()
    var out_ptr = result._data.bitcast[Scalar[dtype]]()
    var sum_val: Float32 = 0.0  # Float32 accumulator
    for i in range(numel):
        sum_val += Float32(in_ptr[i])
    out_ptr[0] = Scalar[dtype](sum_val)
```

#### Task 2.2: Update Mean Operations
Apply same pattern to mean operations in reduction.mojo.

### Phase 3: Test Updates

#### Task 3.1: Enable test_fc3_forward_float16
**File**: `/home/mvillmow/ml-odyssey/tests/models/test_lenet5_layers.mojo`
**Lines**: 816-819

Change from:
```mojo
print("  test_fc3_forward_float16... FIXME(#2703)")
```
To:
```mojo
print("  test_fc3_forward_float16...", end="")
test_fc3_forward_float16()
print(" OK")
```

#### Task 3.2: Add Numerical Accuracy Test
**File**: `/home/mvillmow/ml-odyssey/tests/models/test_lenet5_layers.mojo` (new test)

Add a test that validates numerical accuracy, not just absence of NaN/Inf:

```mojo
fn test_fc3_forward_float16_accuracy() raises:
    """Test FC3 float16 output accuracy against float32 reference."""
    # Compute in float32 (reference)
    var dtype_fp32 = DType.float32
    var _result_fp32 = create_fc3_parameters(dtype_fp32)
    var weights_fp32 = _result_fp32[0]
    var bias_fp32 = _result_fp32[1]
    var input_fp32 = create_special_value_tensor([1, 84], dtype_fp32, SPECIAL_VALUE_ONE)
    var output_fp32 = linear(input_fp32, weights_fp32, bias_fp32)
    
    # Compute in float16
    var dtype_fp16 = DType.float16
    var _result_fp16 = create_fc3_parameters(dtype_fp16)
    var weights_fp16 = _result_fp16[0]
    var bias_fp16 = _result_fp16[1]
    var input_fp16 = create_special_value_tensor([1, 84], dtype_fp16, SPECIAL_VALUE_ONE)
    var output_fp16 = linear(input_fp16, weights_fp16, bias_fp16)
    
    # Compare with tolerance (allow ~1% relative error)
    for i in range(output_fp16.numel()):
        var val_fp16 = output_fp16._get_float64(i)
        var val_fp32 = output_fp32._get_float64(i)
        var rel_error = abs(val_fp16 - val_fp32) / max(abs(val_fp32), 1e-8)
        assert_true(rel_error < 0.01, "FC3 float16 accuracy exceeds 1% tolerance")
```

#### Task 3.3: Add Test to Main Runner
**File**: `/home/mvillmow/ml-odyssey/tests/models/test_lenet5_layers.mojo`
**Lines**: Around line 820

Add the accuracy test call to the main test runner.

### Phase 4: Documentation

#### Task 4.1: Create ADR for Mixed-Precision Accumulation
**File**: `/home/mvillmow/ml-odyssey/docs/adr/ADR-XXX-mixed-precision-accumulation.md` (new file)

Document the architectural decision:
- Problem: Float16 accumulation precision loss in large dot products
- Decision: Use float32 accumulation for float16/bfloat16 operations
- Alternatives considered: Kahan summation, accepting limitations
- Rationale: Industry standard, minimal overhead, maximum compatibility

#### Task 4.2: Update Comments in matrix.mojo
**File**: `/home/mvillmow/ml-odyssey/shared/core/matrix.mojo`

Add module-level documentation explaining the mixed-precision strategy.

### Phase 5: Optional - Kahan Summation (Future Enhancement)

If even higher precision is needed in the future:

#### Task 5.1: Implement Kahan Summation Helper
**File**: `/home/mvillmow/ml-odyssey/shared/core/numerical.mojo` (new file)

```mojo
fn kahan_sum[dtype: DType](values: UnsafePointer[Scalar[dtype]], n: Int) -> Float64:
    """Kahan summation for reduced numerical error."""
    var sum: Float64 = 0.0
    var c: Float64 = 0.0  # Compensation for lost precision
    for i in range(n):
        var y = Float64(values[i]) - c
        var t = sum + y
        c = (t - sum) - y
        sum = t
    return sum
```

#### Task 5.2: Add Kahan Option to Matmul
Create variants that use Kahan summation for extreme precision requirements.

## Files to Modify/Create

| File | Action | Description |
|------|--------|-------------|
| `shared/core/matrix.mojo` | Modify | Add float32 accumulation to all matmul variants |
| `shared/core/reduction.mojo` | Modify | Add float32 accumulation to sum/mean operations |
| `tests/models/test_lenet5_layers.mojo` | Modify | Enable FC3 test, add accuracy test |
| `docs/adr/ADR-XXX-mixed-precision-accumulation.md` | Create | Document architectural decision |
| `shared/core/numerical.mojo` | Create (optional) | Kahan summation utilities |

## Testing Approach

### Unit Tests

1. **Existing test_fc3_forward_float16** - Enable and verify passes
2. **New accuracy test** - Verify float16 results match float32 within 1%
3. **Test all FC layers** - Ensure FC1, FC2, FC3 all pass with float16
4. **Regression tests** - Ensure float32/float64 paths unchanged

### Integration Tests

1. **Full LeNet-5 forward pass** - Run test_all_layers_sequence_float32
2. **End-to-end with float16** - Create E2E test with float16 precision

### Performance Tests

1. **Benchmark matmul** - Ensure no significant regression from float32 accumulation
2. **Memory usage** - Verify no memory impact (accumulator is register-only)

## Success Criteria

1. **test_fc3_forward_float16 passes** - No NaN/Inf, correct shape/dtype
2. **Numerical accuracy** - Float16 output within 1% of float32 reference
3. **No regressions** - All existing tests continue to pass
4. **Performance** - No more than 5% slowdown vs current implementation
5. **Documentation** - ADR created and matrix.mojo comments updated

## Implementation Order

1. **Phase 1** (Core Changes) - Must be done first
2. **Phase 3** (Test Updates) - Can be done in parallel with Phase 1
3. **Phase 2** (Reduction Operations) - After Phase 1 validates approach
4. **Phase 4** (Documentation) - After all code changes complete
5. **Phase 5** (Kahan) - Future enhancement, not required for issue closure

## Risks and Mitigations

| Risk | Probability | Impact | Mitigation |
|------|-------------|--------|------------|
| Performance regression | Low | Medium | Benchmark before/after, optimize if needed |
| Breaks existing tests | Low | High | Run full test suite before merge |
| Float32 not available on all platforms | Very Low | High | Use compile-time checks for dtype support |
| Kahan summation complexity | Medium | Low | Make it optional, document usage |

## Resource Usage
- Tool calls used: 17/50
- Steps used: 17/50
